paymentsheet/src/main/java/com/stripe/android/paymentsheet/flowcontroller/DefaultFlowController.kt

```diff
@@ -424,66 +471,33 @@ internal class DefaultFlowController @Inject internal constructor(
 
     @JvmSynthetic
     internal fun onPaymentOptionResult(
-        paymentOptionResult: PaymentOptionResult?
+        result: PaymentOptionResult?
     ) {
-        paymentOptionResult?.paymentMethods?.let {
+        result?.paymentMethods?.let {
             val currentState = viewModel.state
 
             viewModel.state = currentState?.copyPaymentSheetState(
                 customer = currentState.paymentSheetState.customer?.copy(paymentMethods = it)
             )
         }
-        when (paymentOptionResult) {
-            is PaymentOptionResult.Succeeded -> {
-                val paymentSelection = paymentOptionResult.paymentSelection
-                paymentSelection.hasAcknowledgedSepaMandate = true
-                viewModel.paymentSelection = paymentSelection
-                if (paymentSelection is Link) {
-                    paymentSelection.linkAccount?.let { updateStateWithLinkAccount(it) }
-                }
-                paymentOptionCallback.onPaymentOption(
-                    paymentOptionFactory.create(
-                        paymentSelection
-                    )
-                )
-            }
-            is PaymentOptionResult.Failed -> {
-                paymentOptionCallback.onPaymentOption(
-                    viewModel.paymentSelection?.let {
-                        paymentOptionFactory.create(it)
-                    }
-                )
-            }
-            is PaymentOptionResult.Canceled -> {
-                val paymentSelection = paymentOptionResult.paymentSelection
-                viewModel.paymentSelection = paymentSelection
-
-                val paymentOption = paymentSelection?.let { paymentOptionFactory.create(it) }
-                paymentOptionCallback.onPaymentOption(paymentOption)
-            }
-            null -> {
-                viewModel.paymentSelection = null
-                paymentOptionCallback.onPaymentOption(null)
-            }
+        when (result) {
+            is PaymentOptionResult.Succeeded -> setAsSelected(
+                result.paymentSelection.also { it.hasAcknowledgedSepaMandate = true }
+            )
+            is PaymentOptionResult.Failed -> setAsSelected(viewModel.paymentSelection)
+            is PaymentOptionResult.Canceled -> setAsSelected(result.paymentSelection)
+            null -> setAsSelected(null)
```

Comment:
extracted common logic to `setAsSelected`. Also ensures we extract the LinkAccount from the current payment selection and save it on the FlowController state.

Reply:
Is this a behavior change in the `Failed` case, where we currently don't update `paymentSelection`?